### PR TITLE
[resourcemanager] Parse all node config JSON fields

### PR DESCRIPTION
**Pull request by mlohvynenko**

### CODE REVIEW

File: src/resourcemanager/resourcemanager.cpp [hunk 1]
```diff
@@ -16,21 +16,282 @@ namespace aos::zephyr::resourcemanager {
...
+ * Private
+ **********************************************************************************************************************/
+
+static void FillJSONStruct(const Array<DeviceInfo>& inDevices, NodeConfig& jsonNodeConfig)
+{
+    jsonNodeConfig.devicesLen = inDevices.Size();
+
+    for (size_t i = 0; i < inDevices.Size(); ++i) {
+        const auto& inDevice   = inDevices[i];
+        auto&       jsonDevice = jsonNodeConfig.devices[i];
```
Comment (MykolaSolyanko) [inline src/resourcemanager/resourcemanager.cpp:143]: should we check is `inDevices.Size() <= len jsonNodeConfig.devices` ?

File: src/resourcemanager/resourcemanager.cpp [hunk 2]
```diff
@@ -16,21 +16,282 @@ namespace aos::zephyr::resourcemanager {
...
+
+        jsonDevice.name           = inDevice.mName.CStr();
+        jsonDevice.sharedCount    = inDevice.mSharedCount;
+        jsonDevice.groupsLen      = inDevice.mGroups.Size();
+        jsonDevice.hostDevicesLen = inDevice.mHostDevices.Size();
+
+        for (size_t j = 0; j < inDevice.mGroups.Size(); ++j) {
+            jsonDevice.groups[j] = inDevice.mGroups[j].CStr();
```
Comment (MykolaSolyanko) [inline src/resourcemanager/resourcemanager.cpp:151]: ditto

File: src/resourcemanager/resourcemanager.cpp [hunk 3]
```diff
@@ -16,21 +16,282 @@ namespace aos::zephyr::resourcemanager {
...
+        }
+
+        for (size_t j = 0; j < inDevice.mHostDevices.Size(); ++j) {
+            jsonDevice.hostDevices[j] = inDevice.mHostDevices[j].CStr();
```
Comment (MykolaSolyanko) [inline src/resourcemanager/resourcemanager.cpp:155]: ditto

File: src/resourcemanager/resourcemanager.cpp [hunk 4]
```diff
@@ -16,21 +16,282 @@ namespace aos::zephyr::resourcemanager {
...
+    }
+}
+
+static void FillJSONStruct(const Array<ResourceInfo>& inResources, NodeConfig& jsonNodeConfig)
+{
+    jsonNodeConfig.resourcesLen = inResources.Size();
+
+    for (size_t i = 0; i < inResources.Size(); ++i) {
+        const auto& inResource   = inResources[i];
+        auto&       jsonResource = jsonNodeConfig.resources[i];
```
Comment (MykolaSolyanko) [inline src/resourcemanager/resourcemanager.cpp:166]: ditto

**Review by mykola-kobets-epam — APPROVED** [review]

File: src/resourcemanager/resourcemanager.cpp (hunk 1 above)
Comment (mlohvynenko) [reply src/resourcemanager/resourcemanager.cpp:143]: Didn't get the point.
Do you mean array max size checks? (for json decode into aos it's done via Array::Resize calls, FillAosStruct for example).

For json encode from aos, no check are done. But anyway our structs rely on aos consts, as well as JSON object desriptors that are declared static in this cpp.

BTW, ocispec doesn't check boundaries when aos is converted into JSON as well

**Review by MykolaSolyanko — APPROVED** [review]

File: src/resourcemanager/resourcemanager.cpp [hunk 5]
```diff
@@ -71,6 +345,16 @@ Error JSONProvider::ParseNodeConfig(const String& json, sm::resourcemanager::Nod
 
     auto parsedNodeConfig = MakeUnique<NodeConfig>(&mAllocator);
 
+    memset(parsedNodeConfig.Get(), 0, sizeof(NodeConfig));
```
Comment (al1img) [inline src/resourcemanager/resourcemanager.cpp:348]: ```parsedNodeConfig = {};```

Comment (mlohvynenko) [reply src/resourcemanager/resourcemanager.cpp]: Looks like we don't need this memset at all
All the fields are explicitly initialized below.

BTW, parsedNodeConfig has all it fields set to zero after MakeUnique. (checked with gdb)

**Review by al1img — APPROVED** [review]

**Pull request merged**